### ISSUE TITLE
Inconsistent memory allocation in mutt_option_get()

### ISSUE THREAD

**Issue by guyzmo** [neomutt/neomutt#588]
fix the function so it takes a `struct option_t*` parameter and fill it up with the requested option. Fixing the inconsistency between they `my_` variables path and the general case. cf [coverity check](https://scan9.coverity.com/reports.htm#v32150/p12145/fileInstanceId=16983354&defectInstanceId=3296734&mergedDefectId=169546&fileStart=1&fileEnd=250) (and the ones related to that one).

so new signature would be:

```c
struct option_t *mutt_option_get(const char *, struct option_t *);
```

**Comment by gahr** [comment]
Wouldn't move the `safe_malloc` after the checks be enough? Why the need for the change in the API?

**Comment by gahr** [comment]
How about this?
```diff
diff --git a/init.c b/init.c
index b1e3dd9f..4eabb9c2 100644
--- a/init.c
+++ b/init.c
@@ -256,19 +256,16 @@ const struct option_t *mutt_option_get(const char *s)
   int idx = mutt_option_index(s);
   if (idx != -1)
     return &MuttVars[idx];
-  else if (mutt_strncmp("my_", s, 3) == 0)
+  if (mutt_strncmp("my_", s, 3) == 0)
   {
-    struct option_t *opt = safe_malloc(sizeof(struct option_t));
-    if (!myvar_get(s))
+    const char * myv = myvar_get(s);
+    if (!myv || !*myv)
       return NULL;
-    opt->data = (unsigned long) safe_strdup(myvar_get(s));
-    if (*((char **) opt->data))
-    {
-      opt->option = safe_strdup(s);
-      opt->type = DT_STR;
-      return opt;
-    }
-    FREE(&opt);
+    struct option_t *opt = safe_malloc(sizeof(struct option_t));
+    opt->data = (unsigned long) safe_strdup(myv);
+    opt->option = safe_strdup(s);
+    opt->type = DT_STR;
+    return opt;
   }
   return NULL;
 }
```

**Comment by guyzmo** [comment]
that's one of two options.

Because of the way we handle buffer/token structs in other parts of the 
code, where the caller handles allocation, I thought it'd be more 
consistent to change the prototype to accept an empty `option_t` and fill 
it up.

All in all, we need to change things in several part of the code, either
to `FREE()` the result of `mutt_option_get()` or to feed it some
allocated memory.

As long as we fix this, I'm fine with both solutions. Any strong
argument for/against one of the options? Or shall we ask Mr @flatcap?

**Comment by gahr** [comment]
Well, I though there would have been more users of the function, thus my initial slight concern in changing the API. In fact, I see exactly two calls to `mutt_options_get`, in `_lua_mutt_get` and `_lua_mutt_set`. The first doesn't free it, the latter does something fishy by allocating two `struct option_t` objects , copying the first into the second, and at the end free'ing the second (but not the first).

There are pros and cons on both sides: if the function returns an allocated struct, you gain in lines of code (no need to set it up at the caller). On the other hand, passing it in from the caller means you can allocate it on the stack, so it's faster. 

I think I remember a discussion about modernizing other parts of the code to use the newly added functions. Since you'll need to review that code anyway, I think it's up to you to chose the option you prefer :)

**Comment by flatcap** [comment]
Fixed in 5bfcc3b, eef5852.  I opted for the "pass in an empty struct" method.
The diffs look big, but most of the changes are things like `opt->` to `opt.`

Whether we're getting a variable or a my_var, we're returning pointers to existing strings.
Nothing is being allocated.

If we allocated a struct then we'd have to tell the caller: free the struct but not its contents.

The second commit was changing some of the `Buffer` usage.  The `Buffer` structs we being allocated but not given a string buffer inside.  Some of the error paths were assuming that the string buffers existed.
